Fakify params and weights under private config

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90417

Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that @Chillee  and @bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

tugsbayasgalan committed:
Fakify params and weights under private config

[ghstack-poisoned]

zhxchen17 commented:
Yeah I think this will be helpful to unblock all our use cases, including ASR and DPE models, since all models have weights. Right now it's nonscalable on our side to unlift params and buffers, because we have to handle so many corner cases.

ezyang commented:
I think I've mentioned (to @eellison ? @voznesenskym ?) that I'm open to the idea of automatically fakeifying real tensors when in fake tensor mode in general, and then we would not have to guard this under a config flag. To understand if this is safe or not, we have to understand why we didn't add support for this initially.

Consider the following program:

```
x = torch.empty(0)
with FakeTensorMode():
    x.resize_(10)
    y = x * 2
```

What exactly is this program supposed to do? Naively, we would perform a resize_ on the original tensor x, but this is problematic, because the entire point of fake tensor mode is to avoid doing real compute, and a x.resize_(10) would cause real memory to be allocated. However, nor can we *not* do a resize_ on the original tensor x, because the subsequent computation on x needs the updated metadata from the resize call.

It sounds like you don't need this case. I think we'd be OK with throwing if this happens. Can we do this, instead of guarding this in private config?

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -26,6 +26,18 @@
 
 CONSTANT_NUMEL_LIMIT = 1
 
+CONSTANT_LIFTING = False
```
this should be a property of the FakeTensorMode, not a global property. Maybe `convert_non_fake_inputs`.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -784,9 +796,15 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                    "Invoking operators with non-Fake Tensor inputs in FakeTensorMode is not yet supported. "
+                    f"Please convert all Tensors to FakeTensors first. Found in {func}(*{args}, **{kwargs})"
+                )
+            args, kwargs = tree_map_only(
```
as @ezyang commented, I think we need to keep track of the inputs we lift, and throw if we call torch.Tag.inplace_view` on them.

```
x = torch.rand([4, 4])
with FakeTensorMode(allow_non_fake_inputs=True):
     y = x.add_(3)
     y.resize_([16])
     assert(x.shape == y.shape) # would fail
```

eellison commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -652,6 +652,8 @@ def wrapped(*args):
         if tracing_mode == "symbolic":
             python_dispatcher_mode = enable_python_dispatcher()
 
+        constant_lifting_mode = _enable_constant_lifting() if _constant_lifting else nullcontext()
```
constant_lifting is a bit of a misnomer at the moment. we don't have any guards in place to ensure that the arguments we lift are constant.

tugsbayasgalan replied:
Sure I can rename and put this under FakeTensorMode

eellison reviewed (commented):
I think checking for inplace views is sufficient here, but just thinking aloud, are there any other cases where someone would use `FakeTensorMode(allow_non_fake_inps=True)` and unwittingly affect real state ? Maybe by accessing the `param.grad` ?

eellison commented:
We should also consider the api for if the user will want to make the lifted inputs static or dynamic

tugsbayasgalan commented:
> I think checking for inplace views is sufficient here



> I think I've mentioned (to @eellison ? @voznesenskym ?) that I'm open to the idea of automatically fakeifying real tensors when in fake tensor mode in general, and then we would not have to guard this under a config flag. To understand if this is safe or not, we have to understand why we didn't add support for this initially.
> 
> Consider the following program:
> 
> ```
> x = torch.empty(0)
> with FakeTensorMode():
>     x.resize_(10)
>     y = x * 2
> ```
> 
> What exactly is this program supposed to do? Naively, we would perform a resize_ on the original tensor x, but this is problematic, because the entire point of fake tensor mode is to avoid doing real compute, and a x.resize_(10) would cause real memory to be allocated. However, nor can we _not_ do a resize_ on the original tensor x, because the subsequent computation on x needs the updated metadata from the resize call.
> 
> It sounds like you don't need this case. I think we'd be OK with throwing if this happens. Can we do this, instead of guarding this in private config?

Are you referring to mutation on global variable? I think local mutations on the inputs should be handled by functionalization right?

ezyang commented:
Mutation on a global variable, but in this case it refers to parameters which you didn't fakeify.

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change. 


[ghstack-poisoned]

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -784,9 +789,15 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+            if (torch.Tag.inplace_view in func.tags) or (not self.allow_non_fake_inputs):
+                raise Exception(
+                    "Invoking operators with non-Fake Tensor inputs in FakeTensorMode is not yet supported. "
+                    f"Please convert all Tensors to FakeTensors first. Found in {func}(*{args}, **{kwargs})"
```
Can we update the error message now that this is supported and to suggest `allow_non_fake_inputs` ? And have a separate error msg for the `torch.Tag.inplace_view` case

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change. 


[ghstack-poisoned]

ezyang commented:
i'm going to let @eellison drive the review, holler if you need second opinion

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -720,6 +725,8 @@ def __init__(
 
         self.shape_env = shape_env
 
+        self.fakified_real_tensors = set()
```
this is unused, plz delete. 
```
with FakeTensorMode(allow_non_fake_inputs=True):
     y = x.add_(3)
     print(isinstance(y, FakeTensor) # False
```

The reason we don't need it is because aten codegen returns the input instead of the output here.

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -783,11 +790,9 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-                "Invoking operators with non-Fake Tensor inputs in FakeTensorMode is not yet supported. "
-                f"Please convert all Tensors to FakeTensors first. Found in {func}(*{args}, **{kwargs})"
-            )
+        args, kwargs = self.validate_and_convert_non_fake_tensors(
```
could we either `del flat_arg_tensors` after the call here or update it to be correct ? if someone were to add another use of `flat_arg_tensors` later it would be outdated

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -910,10 +915,29 @@ def check_for_subclass(self, flat_arg_tensors):
+                        f"Can't call inplace view ops on global variables or model attributes, Found in {func}(*{args}, **{kwargs})"
+                    )
+                if not self.allow_non_fake_inputs:
+                    raise Exception(
```
update the error message to reflect the new flag here.

`f"Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. 
 Found in {func}(*{args}, **{kwargs}) "`

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -910,10 +915,29 @@ def check_for_subclass(self, flat_arg_tensors):
+            if isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor):
+                if torch.Tag.inplace_view in func.tags:  # type: ignore[attr-defined]
+                    raise Exception(
+                        f"Can't call inplace view ops on global variables or model attributes, Found in {func}(*{args}, **{kwargs})"
```
nit: maybe `Can't call metadata mutating ops on non-Fake Tensor Inputs.` ? inplace view is not a user-facing concept, and we might run with non globals/model attirbutes.

eellison reviewed (approved):
LGTM! just a few remaining small nits

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change. 


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change. 


[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Fakify params and weights under private config"


Previously, we planned to lift the parameters and weights while exporting and implement our own transformer to "unlift" the lifted weights and params back to the graph as attributes. But this is bit challenging because:

- We need to maintain correct ordering for weights and parameters that are passed as inputs so that we know how to map them back. 
- Some weights are unused in the graph, so our transformer needs to be aware of which weights and parameters are not used in the graph. And we need to distinguish which are real user input and which are parameters. 
- There can be more edge cases we haven't seen in other models yet. 

I am aware that Chillee  and bdhirsh mentioned that functionalization won't work with fake-tensor attributes but this is fine for the short term as we don't expect users to be modifying weights and params in inference mode. In fact, we explicitly disable attribute mutation in torchdynamo export mode right now. 

Given above condition, it might be ok to just fakify params when we need. I use a flag to guard against this change.

Differential Revision: [D41891201](https://our.internmc.facebook.com/intern/diff/D41891201)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41891201).

tugsbayasgalan commented:
@pytorchbot merge
